src/blueapi/service/authentication.py

```diff
@@ -0,0 +1,218 @@
+from __future__ import annotations
+
+import base64
+import json
+import os
+import time
+from abc import ABC, abstractmethod
+from enum import Enum
+from http import HTTPStatus
+from pathlib import Path
+from typing import Any
+
+import jwt
+import requests
+
+from blueapi.config import (
+    CLIClientConfig,
+    OAuthClientConfig,
+    OAuthServerConfig,
+)
+
+
+class AuthenticationType(Enum):
+    DEVICE = "device"
+    PKCE = "pkce"
+
+
+class Authenticator:
+    def __init__(
+        self,
+        server_config: OAuthServerConfig,
+        client_config: OAuthClientConfig,
+    ):
+        self._server_config: OAuthServerConfig = server_config
+        self._client_config: OAuthClientConfig = client_config
+
+    def decode_jwt(self, token: str, verify_expiration: bool = True) -> dict[str, str]:
+        signing_key = jwt.PyJWKClient(
+            self._server_config.jwks_uri
+        ).get_signing_key_from_jwt(token)
+        decode: dict[str, str] = jwt.decode(
+            token,
+            signing_key.key,
+            algorithms=["RS256"],
+            options={"verify_exp": verify_expiration},
+            verify=True,
+            audience=self._client_config.client_audience,
+            issuer=self._server_config.issuer,
+            leeway=5,
```

Comment:
What is `leeway`?

Reply:
leeway ([float](https://docs.python.org/3/library/functions.html#float)) – a time margin in seconds for the expiration check